R/simulate.r

```diff
@@ -1,3 +1,309 @@
+    offspring_dist,
+    ...,
+    stat_max = Inf,
+    pop = Inf,
+    percent_immune = 0,
+    generation_time = NULL,
+    t0 = 0,
+    tf = Inf) {
+  # Input checking
+  checkmate::assert_count(index_cases, positive = TRUE)
+  checkmate::assert_choice(statistic, choices = c("size", "length"))
+  checkmate::assert_string(offspring_dist)
+  # check that offspring function exists in the environment
+  roffspring_name <- paste0("r", offspring_dist)
+  check_offspring_func_valid(roffspring_name)
+  checkmate::assert(
+    is.infinite(stat_max) ||
+      checkmate::assert_integerish(stat_max, lower = 0)
+  )
+  checkmate::assert(
+    is.infinite(pop) ||
+      checkmate::assert_integerish(pop, lower = 0)
+  )
+  checkmate::assert_number(
+    percent_immune,
+    lower = 0, upper = 1
+  )
+  if (!missing(generation_time)) {
+    check_generation_time_valid(generation_time)
+  } else if (!missing(tf)) {
+    stop("If `tf` is specified, `generation_time` must be specified too.")
+  }
+  checkmate::assert_numeric(
+    t0,
+    lower = 0, finite = TRUE
+  )
+  checkmate::assert_number(
+    tf,
+    lower = 0
+  )
+  # Gather offspring distribution parameters
+  pars <- list(...)
+
+  # Initialisations
+  stat_track <- rep(1, index_cases) # track statistic (length or size)
+  n_offspring <- rep(1, index_cases) # current number of offspring
+  sim <- seq_len(index_cases) # track trees that are still being simulated
+  infector_ids <- rep(1, index_cases)
+
+  # initialise list of data frame to hold the transmission trees
+  generation <- 1L
+  tree_df <- list(
+    data.frame(
+      infectee_id = seq_len(index_cases),
+      sim_id = 1L,
+      infector_id = NA_integer_,
+      generation = generation
+    )
+  )
+  # Initialise susceptible population
+  # I think this currently implicitly assumes the initial cases are imported
+  # (not subtracted from pop). If not, we'll have to subtract the non-imported
+  # cases. An improvement of this could take the form of an extra argument that
+  # defines a percentage of initial cases suspected to be imported. We can
+  # then substract only the non-imported cases from pop.
+  susc_pop <- round(pop * (1 - percent_immune))
+
+  # Add optional columns
+  if (!missing(generation_time)) {
+    tree_df[[generation]]$time <- t0
+    times <- tree_df[[generation]]$time
+  }
+  if (!missing(pop)) {
+    tree_df[[generation]]$susc_pop <- susc_pop
+  }
+  # next, simulate n trees
+  while (length(sim) > 0 && susc_pop > 0) {
+    # simulate next generation
+    next_gen <- do.call(
+      roffspring_name,
+      c(
+        list(n = sum(n_offspring[sim])),
+        pars
+      )
+    )
+    if (any(next_gen %% 1 > 0)) {
+      stop("Offspring distribution must return integers")
+    }
+    # Adjust next_gen if sum exceeds the number of susceptibles
+    if (sum(next_gen) > susc_pop) {
+      scaling_factor <- susc_pop / sum(next_gen)
+      next_gen <- floor(next_gen * scaling_factor)
```

Comment:
it might be preferable to have something that is guaranteed to set `sum(next_gen)` to `sum_pop` and is guaranteed not to leave any susceptibles. I think that could be achieved with

```suggestion
      ## create hypothetical next generation population to sample from
      next_gen_pop <- rep(seq_along(next_gen), times = next_gen)
      ## sample from hypothetical next generation
      next_gen_sample <- sample(x = next_gen_pop, size = susc_pop)
      ## create adjusted next_gen vector
      next_gen <- rep(0L, length(next_gen))
      ## count occurrences in next generation sample
      next_gen_count <- table(next_gen_sample)
      next_gen[as.integer(names(next_gen_count))] <- unname(next_gen_count)
```